### PR TITLE
fix: deduplicate FormattedCommunicationEndpoints in ConnectionInfo - ICP-3025

### PR DESCRIPTION
**Pull request by VLazarevic**
## Description

https://dt-rnd.atlassian.net/browse/ICP-3025

This is a follow-up of the discussion in PR #6371.

Before, the `CommunicationEndpoints []string` field was used to de-duplicate the items. Now only the `FormattedCommunicationEndpoints string` field is used, so the old de-duplication logic does not apply anymore. Because of this it was possible that `GetConnectionInfo` returns the same endpoint multiple times.

This PR adds de-duplication for `FormattedCommunicationEndpoints` so that `GetConnectionInfo` returns only unique endpoints. For the parsing/de-duplication I oriented myself on the existing logic that is already used for istio in `pkg/controllers/dynakube/istio/communication_hosts.go`, to stay consistent with the rest of the code.

## How can this be tested?

There are unit tests added that cover the de-duplication function directly (no duplicates, some duplicates, all duplicates, empty input, single endpoint). You can run them with:

```
go test ./pkg/clients/dynatrace/...
```

I also tested it manually in a cluster:
1. I deployed a small mock server inside the cluster that returns a `connectioninfo` response where the same endpoint is contained multiple times in `formattedCommunicationEndpoints` (e.g. `https://endpoint1...;https://endpoint1...;https://endpoint2...`).
2. I applied a DynaKube with `oneAgent.classicFullStack` and pointed the `apiUrl` to the mock server.
3. After the reconcile I checked the DynaKube status with `kubectl -n dynatrace get dynakube dynakube -o yaml`.

In the status under `oneAgent.connectionInfoStatus.endpoints` the duplicated endpoint appears only one time now, so the de-duplication works as expected:

```
endpoints: https://endpoint1.dynatrace.com:443;https://endpoint2.dynatrace.com:443
```

### CODE REVIEW

File: pkg/clients/dynatrace/oneagent/connection_info.go [hunk 1]
```diff
@@ -50,5 +56,34 @@ func (c *ClientImpl) GetConnectionInfo(ctx context.Context) (ConnectionInfo, err
...
 	}
 
+	resp.Endpoints = deduplicateEndpoints(resp.Endpoints)
+
 	return resp, nil
 }
+
+// deduplicateEndpoints removes duplicate endpoints from a formatted
+// communication-endpoints string (delimiter-separated, see endpointDelimiter).
+func deduplicateEndpoints(endpoints string) string {
```
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info.go]: The comment in the connection info struct was targeted at the implementer of this ticket. The API already gives us a list, so we can use it here.

Comment (jokuniew) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: @avorima so should we let `FormattedCommunicationEndpoints` go?
Asking cause I've might formulate the ticket incorrectly

Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: So I adapted the struct, see [d150ef2](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/d150ef2c6729887c94b89a271f4cfddf2b1afd35). Please have a look if that way works for you :)

Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: > @avorima so should we let `FormattedCommunicationEndpoints` go? Asking cause I've might formulate the ticket incorrectly

If we let `FormattedCommunicationEndpoints` go, then we need to adapt the tests and remove the duplicates in `CommunicationEndpoints` as well. Is this our goal? IMO I wouldn't remove it 🤔

File: pkg/clients/dynatrace/oneagent/connection_info.go [hunk 2]
```diff
@@ -13,15 +14,18 @@ import (
 
 const (
 	connectionInfoPath = "/v1/deployment/installer/agent/connectioninfo"
+
+	// endpointDelimiter separates the individual endpoints inside the
+	// formattedCommunicationEndpoints string returned by the connectioninfo API.
+	// It matches the delimiter used by connectioninfo.ParseOACommunicationHosts.
+	endpointDelimiter = ";"
```
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info.go]: consumers of the client packages should not have to deal with the weirdness of the API. use the standard separator and get rid of our special handling

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 1 above)
Comment (avorima) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: the goal is that none of the consumers of this client package deal with duplicate endpoints

Comment (jokuniew) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: but letting go I mean to not unmarshall the FormattedCommunicationEndpoints from response
```
type ConnectionInfo struct {
	TenantUUID             string   `json:"tenantUUID"`
	TenantToken            string   `json:"tenantToken"`
	Endpoints              string   # here
	CommunicationEndpoints []string `json:"communicationEndpoints"`
}
```

Comment (jokuniew) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: @avorima to be super correct, shouldn't we then create private struct that will fetch `CommunicationEndpoints []string json:"communicationEndpoints"` and return from this method a public struct with `Endpoints              string ` then? So we'll have dedup guaranteed and user will not use CommunicationEndpoints array accidentally

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 2 above)
Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: [9df3582](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/9df3582099926b9b08e3ad3ae730aac1ef9e9b43)

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 1 above)
Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: [08ba018](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/08ba0189ec083b7dbd289d4bc4d88215a3699727) I implemented it like that, was that the idea? @jokuniew

File: pkg/controllers/dynakube/connectioninfo/communication_hosts.go [hunk 3]
```diff
@@ -64,15 +64,9 @@ func NewCommunicationHost(endpoint string) (CommunicationHost, error) {
 	}, nil
 }
 
-// ParseOACommunicationHosts creates CommunicationHost slice from `;` separated endpoints.
+// ParseCommunicationHosts creates CommunicationHost slice from `,` separated endpoints.
 // It removes duplicates and sorts the result for deterministic output.
```
Comment (avorima) [inline pkg/controllers/dynakube/connectioninfo/communication_hosts.go]: This "removes duplicates" isn't accurate anymore

File: pkg/clients/dynatrace/oneagent/connection_info_test.go [hunk 4]
```diff
@@ -107,4 +136,137 @@ func Test_GetConnectionInfo(t *testing.T) {
 		_, err := oaClient.GetConnectionInfo(ctx)
 		assert.ErrorIs(t, err, expectErr)
 	})
+
+	t.Run("duplicate endpoints are deduplicated", func(t *testing.T) {
```
Comment (jokuniew) [inline pkg/clients/dynatrace/oneagent/connection_info_test.go]: this test is identical as `t.Run("endpoints are derived from the deduplicated communicationEndpoints slice", func(t *testing.T) {
` I think it can be removed

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 1 above)
Comment (jokuniew) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: I think yes. @avorima ?

File: pkg/clients/dynatrace/oneagent/connection_info.go [hunk 5]
```diff
@@ -50,5 +58,34 @@ func (c *ClientImpl) GetConnectionInfo(ctx context.Context) (ConnectionInfo, err
...
+	}, nil
+}
+
+// deduplicateEndpoints removes duplicate entries from the communication
+// endpoints returned by the API and joins the unique ones into a single,
+// comma-separated string. The order of first occurrence is preserved,
+// surrounding whitespace is trimmed and empty entries are dropped.
+func deduplicateEndpoints(endpoints []string) string {
+	seen := make(map[string]struct{})
+	unique := make([]string, 0)
```
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info.go]: prefer use of `var slice []string` or `slice := []string{}` if you're not preallocating

File: pkg/controllers/dynakube/connectioninfo/communication_hosts_test.go [hunk 6]
```diff
@@ -245,7 +245,7 @@ func TestParseAGCommunicationHosts(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			hosts, err := ParseAGCommunicationHosts(tc.input)
+			hosts, err := ParseCommunicationHosts(tc.input)
```
Comment (avorima) [inline pkg/controllers/dynakube/connectioninfo/communication_hosts_test.go:166]: these tests are now duplicated

File: pkg/clients/dynatrace/oneagent/connection_info_test.go [hunk 7]
```diff
@@ -107,4 +136,137 @@ func Test_GetConnectionInfo(t *testing.T) {
...
+		assert.Equal(t, expected, connectionInfo)
+	})
+}
+
+// Test_GetConnectionInfo_EndToEnd exercises the full path through the real core
+// HTTP client against an httptest.Server. The mocked JSON response contains both
+// the formatted string and a communicationEndpoints array with duplicates, and we
+// assert the returned Endpoints string is derived from the slice with duplicates
+// removed and the entries joined by a comma.
+func Test_GetConnectionInfo_EndToEnd(t *testing.T) {
```
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info_test.go]: remove this test. we don't do httpserver in subclient packages

File: pkg/clients/dynatrace/oneagent/connection_info_test.go (hunk 4 above)
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info_test.go]: duplicate test

Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info_test.go]: [e2e3902](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/e2e39029a360d7df45c3c2712b68aec101ed5736)

Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info_test.go]: [e2e3902](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/e2e39029a360d7df45c3c2712b68aec101ed5736)

File: pkg/clients/dynatrace/oneagent/connection_info_test.go (hunk 7 above)
Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info_test.go]: [152c401](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/152c401d18ddf465d13383c6bc464cb0e5e15a54)

File: pkg/controllers/dynakube/connectioninfo/communication_hosts_test.go (hunk 6 above)
Comment (VLazarevic) [reply pkg/controllers/dynakube/connectioninfo/communication_hosts_test.go:166]: [152c401](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/152c401d18ddf465d13383c6bc464cb0e5e15a54)

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 5 above)
Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: [cacb4b2](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/cacb4b2255ded333ce49cb290f6ddf0acfb42717)

File: pkg/controllers/dynakube/connectioninfo/communication_hosts.go (hunk 3 above)
Comment (VLazarevic) [reply pkg/controllers/dynakube/connectioninfo/communication_hosts.go]: [9aa94f9](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/9aa94f904d0dd9d5bf8cdb44cd25d8f9feb865c3)

File: pkg/clients/dynatrace/oneagent/connection_info.go [hunk 8]
```diff
@@ -15,19 +16,26 @@ const (
...
+// connectionInfoResponse is the raw shape returned by the connectioninfo API.
+// It is unmarshalled directly from JSON and never exposed outside this package.
+type connectionInfoResponse struct {
+	TenantUUID             string   `json:"tenantUUID"`
+	TenantToken            string   `json:"tenantToken"`
+	CommunicationEndpoints []string `json:"communicationEndpoints"`
+}
+
+// ConnectionInfo is the public result of GetConnectionInfo. Endpoints is
+// always deduplicated and comma-separated, consumers never see the raw
```
Comment (avorima) [inline pkg/clients/dynatrace/oneagent/connection_info.go:28]: sentence is cutoff

File: pkg/controllers/dynakube/connectioninfo/communication_hosts.go
Comment (avorima) [inline pkg/controllers/dynakube/connectioninfo/communication_hosts.go]: the map is no longer needed

Comment (avorima) [inline pkg/controllers/dynakube/connectioninfo/communication_hosts.go:90]: i would expect two hosts whose `String()` output is equal to also have equal member values. i think this can be simplified

Comment (VLazarevic) [reply pkg/controllers/dynakube/connectioninfo/communication_hosts.go:90]: [bc6d481](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/bc6d481fd1ec1469654b66ad888b01ddd71359bb)

Comment (VLazarevic) [reply pkg/controllers/dynakube/connectioninfo/communication_hosts.go]: [bc6d481](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/bc6d481fd1ec1469654b66ad888b01ddd71359bb)

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 8 above)
Comment (VLazarevic) [reply pkg/clients/dynatrace/oneagent/connection_info.go:28]: [bc6d481](https://github.com/Dynatrace/dynatrace-operator/pull/7195/commits/bc6d481fd1ec1469654b66ad888b01ddd71359bb)

**Review by avorima — APPROVED** [review]

File: pkg/clients/dynatrace/oneagent/connection_info.go (hunk 2 above)
Comment (0sewa0) [reply pkg/clients/dynatrace/oneagent/connection_info.go]: does the OneAgent accept the endpoints when you do this?

Comment (0sewa0) [reply pkg/clients/dynatrace/oneagent/connection_info.go]:
```
026-08-05 10:18:01.111159 UTC [10/0000000a] info    [comm  ] Initial connect: connection to initial gateways failed (last error http status: 400) using any of
2026-08-05 10:18:01.111225 UTC [10/0000000a] info    [comm  ]       {https://34.118.231.113:443/communication,https://marcell-dev-activegate.dynatrace:443/communication,https://deve2e-us-east-1b.dev.dynatracelabs.com/communication,https://deve2e-us-east-1a.dev.dynatracelabs.com/communication,https://deve2e-us-east-1d.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-211-153-171-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-52-5-34-162-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-52-44-29-255-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-164-6-77-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-88-45-104-deve2e.dev.dynatracelabs.com/communication,https://hvs0985d.dev.dynatracelabs.com:443/communication}
2026-08-05 10:18:01.111347 UTC [10/0000000a] warning [native] Unable to do initial setup because no server is reachable. Last error: http status: 400
2026-08-05 10:18:01.111537 UTC [10/0000000a] info    [native] Dispatcher started - using non working URL https://34.118.231.113:443/communication,https://marcell-dev-activegate.dynatrace:443/communication,https://deve2e-us-east-1b.dev.dynatracelabs.com/communication,https://deve2e-us-east-1a.dev.dynatracelabs.com/communication,https://deve2e-us-east-1d.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-211-153-171-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-52-5-34-162-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-52-44-29-255-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-164-6-77-deve2e.dev.dynatracelabs.com/communication,https://sg-us-east-1-54-88-45-104-deve2e.dev.dynatracelabs.com/communication,https://hvs0985d.dev.dynatracelabs.com:443/communication
```
doesn't seem like it...
